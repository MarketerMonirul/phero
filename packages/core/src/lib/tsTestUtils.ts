import ts, {
  FunctionLikeDeclarationBase,
  VariableDeclaration,
} from "typescript"
import { TSFiles, VirtualCompilerHost } from "./VirtualCompilerHost"
import { PheroApp } from "../domain/PheroApp"
import { KindToNodeMappings } from "./tsUtils"
import {
  DependencyMap,
  FunctionParserModel,
  generateParserModel,
  ParserModelMap,
} from "../generate-model-3/generateParserModel"
import { ObjectParserModel, ParserModel } from "../generate-model-3/ParserModel"

export function printPheroApp(app: PheroApp): string {
  return JSON.stringify(
    {
      services: app.services.map((service) => ({
        name: service.name,
        funcs: service.funcs.map((func) => ({
          name: func.name,
          func: printFunctionDeclaration(func.ref),
        })),
      })),
    },
    null,
    4,
  )

  function printFunctionDeclaration(
    func: FunctionLikeDeclarationBase | VariableDeclaration,
  ): string {
    const funcName = func.name?.getText()

    if (!funcName) {
      throw new Error("Func must have name")
    }

    return func.kind === ts.SyntaxKind.FunctionDeclaration
      ? `[FunctionDeclaration(${funcName})]`
      : func.kind === ts.SyntaxKind.VariableDeclaration
      ? `[VariableDeclaration(${funcName})]`
      : func.kind === ts.SyntaxKind.FunctionExpression
      ? `[FunctionExpression(${funcName})]`
      : func.kind === ts.SyntaxKind.ArrowFunction
      ? `[ArrowFunction(${funcName})]`
      : "[UNKNOWN]"
  }
}

export function createTestProgram(input: TSFiles | string): ts.Program {
  const vHost = new VirtualCompilerHost()

  if (typeof input === "string") {
    vHost.addFile(`phero.ts`, input)
  } else {
    for (const [fileName, content] of Object.entries(input)) {
      vHost.addFile(`${fileName}.ts`, content)
    }
  }

  const program = vHost.createProgram("phero.ts")
  return program
}

export function compileStatement<SK extends ts.SyntaxKind>(
  code: string,
  syntaxKind: SK,
): { statement: KindToNodeMappings[SK]; prog: ts.Program } {
  const prog = createTestProgram(code)
  const statements = prog.getSourceFile("phero.ts")?.statements
  if (statements?.length !== 1) {
    throw new Error("Should provide exactly 1 statement")
  }
  const statement = statements[0]
  if (statement.kind !== syntaxKind) {
    throw new Error(
      `SytaxKind of statement is ${statement.kind}, but ${syntaxKind} was expected`,
    )
  }

  return {
    prog,
    statement: statement as KindToNodeMappings[SK],
  }
}

export function compileStatements(code: string): {
  statements: ts.Statement[]
  prog: ts.Program
} {
  const prog = createTestProgram(code)
  const statements = prog.getSourceFile("phero.ts")?.statements
  if (!statements || statements.length < 1) {
    throw new Error("Should provide at least 1 statement")
  }

  return {
    prog,
    statements: statements.map((s) => s),
  }
}

export function compileProgram(input: TSFiles): {
  statements: ts.Statement[]
  prog: ts.Program
} {
  const prog = createTestProgram(input)
  const statements = prog.getSourceFile("phero.ts")?.statements
  if (!statements || statements.length < 1) {
    throw new Error("Should provide at least 1 statement")
  }

  return {
    prog,
    statements: statements.map((s) => s),
  }
}

const printer = ts.createPrinter({
  newLine: ts.NewLineKind.LineFeed,
  noEmitHelpers: true,
  removeComments: true,
  omitTrailingSemicolon: false,
})

export function printCode(code: ts.Node | ts.Node[]): string {
  const sf = ts.createSourceFile(
    "phero-manifest.d.ts",
    "",
    ts.ScriptTarget.ESNext,
    undefined,
    ts.ScriptKind.TS,
  )

  return Array.isArray(code)
    ? printer.printList(
        ts.ListFormat.SourceFileStatements,
        ts.factory.createNodeArray(code),
        sf,
      )
    : printer.printNode(ts.EmitHint.Unspecified, code, sf)
}

export function generateParserModelForReturnType(tsContent: string): {
  root: ParserModel
  deps: Record<string, ParserModel>
} {
  const { statements, prog } = compileStatements(tsContent)

  const func = statements.find((st): st is ts.FunctionDeclaration =>
    ts.isFunctionDeclaration(st),
  )

  if (!func) {
    throw new Error("Ts content doesn't contain any function")
  }

  const funcModel = generateParserModel(func, prog.getTypeChecker(), new Map())
  return {
    root: funcModel.returnType,

    deps: [...funcModel.deps.entries()].reduce<Record<string, ParserModel>>(
      (result, [name, model]) => ({ ...result, [name]: model }),
      {},
    ),
  }
}

export function generateParserModelForFunction(tsContent: string): {
  returnType: ParserModel
  parameters: ObjectParserModel
  deps: Record<string, ParserModel>
} {
  const { statements, prog } = compileStatements(tsContent)

  const func = statements.find((st): st is ts.FunctionDeclaration =>
    ts.isFunctionDeclaration(st),
  )

  if (!func) {
    throw new Error("Ts content doesn't contain any function")
  }

  const funcModel = generateParserModel(func, prog.getTypeChecker(), new Map())

  return {
    returnType: funcModel.returnType,
    parameters: funcModel.parameters,
    deps: [...funcModel.deps.entries()].reduce<Record<string, ParserModel>>(
      (result, [name, model]) => ({ ...result, [name]: model }),
      {},
    ),
  }
}
