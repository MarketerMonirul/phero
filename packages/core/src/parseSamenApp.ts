import ts, { ClassDeclaration } from "typescript"
import { ParseError } from "./errors"
import { ParsedError } from "./extractErrors/parseThrowStatement"
import extractServiceFromSamenExport from "./extractServiceFromSamenExport"
import { hasModifier } from "./tsUtils"

export interface ParsedSamenApp {
  models: Model[]
  errors: ParsedError[]
  services: ParsedSamenServiceDefinition[]
}

export interface ParsedSamenServiceDefinition {
  name: string
  models: Model[]
  errors: ParsedError[]
  funcs: ParsedSamenFunctionDefinition[]
  config: ParsedSamenServiceConfig
}

export type Model =
  | ts.InterfaceDeclaration
  | ts.TypeAliasDeclaration
  | ts.EnumDeclaration

export interface ParsedSamenFunctionDefinition {
  name: string
  // TODO for clashing service models
  // models: Model[]
  actualFunction: ts.FunctionLikeDeclarationBase
  parameters: ts.ParameterDeclaration[]
  returnType: ts.TypeNode
  config: ParsedSamenFunctionConfig
  serviceContext?: {
    type: ts.TypeNode
    paramName?: string
  }
}

export interface ParsedSamenServiceConfig {
  memory?: number
  timeout?: number

  minInstance?: number
  maxInstance?: number
  middleware?: ParsedMiddlewareConfig[]
  contextType?: ts.TypeNode
}

export interface ParsedSamenFunctionConfig {
  memory?: number
  timeout?: number
}

export interface ParsedMiddlewareConfig {
  paramsType: ts.TypeNode
  nextType: ts.TypeNode | undefined
  contextType: ts.TypeNode
  middleware: ts.FunctionLikeDeclarationBase
}

export enum SamenLibFunctions {
  CreateService = "createService",
  CreateFunction = "createFunction",
}

export default function parseSamenApp(
  samenSourceFile: ts.SourceFile,
  typeChecker: ts.TypeChecker,
): ParsedSamenApp {
  const exportStatements = samenSourceFile.statements.filter(
    (s) =>
      hasModifier(s, ts.SyntaxKind.ExportKeyword) || ts.isExportDeclaration(s),
  )

  const services: ParsedSamenServiceDefinition[] = []

  for (const statement of exportStatements) {
    if (ts.isVariableStatement(statement)) {
      for (const varDeclr of statement.declarationList.declarations) {
        const service = extractServiceFromSamenExport(varDeclr, typeChecker)
        services.push(service)
      }
    } else if (ts.isExportDeclaration(statement)) {
      if (!statement.exportClause) {
        throw new ParseError(
          `"export * from './file'" are not supported`,
          statement,
        )
      } else if (!ts.isNamedExports(statement.exportClause)) {
        throw new ParseError("Unsupported export statement", statement)
      }

      for (const specifier of statement.exportClause.elements) {
        const service = extractServiceFromSamenExport(specifier, typeChecker)
        services.push(service)
      }
    } else {
      throw new ParseError("Unsupported export statement", statement)
    }
  }

  const modelMap: Map<Model, string[]> = new Map<Model, string[]>()

  for (const service of services) {
    for (const model of service.models) {
      if (modelMap.has(model)) {
        if (!modelMap.get(model)!.includes(service.name)) {
          modelMap.set(model, [...modelMap.get(model)!, service.name])
        }
      } else {
        modelMap.set(model, [service.name])
      }
    }
  }

  const sharedModels: Model[] = []
  for (const [model, serviceNames] of modelMap) {
    if (serviceNames.length > 1) {
      sharedModels.push(model)
    }
  }

  const errorMap: Map<ClassDeclaration, string[]> = new Map<
    ClassDeclaration,
    string[]
  >()

  for (const service of services) {
    for (const { ref: errorClass } of service.errors) {
      if (errorMap.has(errorClass)) {
        if (!errorMap.get(errorClass)!.includes(service.name)) {
          errorMap.set(errorClass, [...errorMap.get(errorClass)!, service.name])
        }
      } else {
        errorMap.set(errorClass, [service.name])
      }
    }
  }

  const sharedErrors: ParsedError[] = []
  const sharedErrorClasses: ClassDeclaration[] = []

  const allParsedErrors = services.flatMap((s) => s.errors)
  for (const [errorClass, serviceNames] of errorMap) {
    if (serviceNames.length > 1) {
      const parsedErr = allParsedErrors.find((e) => e.ref === errorClass)!

      sharedErrors.push(parsedErr)
      sharedErrorClasses.push(errorClass)
    }
  }

  return {
    models: sharedModels,
    errors: sharedErrors,
    services: services.map((service) => ({
      ...service,
      models: service.models.filter((m) => !sharedModels.includes(m)),
      errors: service.errors.filter((e) => !sharedErrorClasses.includes(e.ref)),
    })),
  }
}
