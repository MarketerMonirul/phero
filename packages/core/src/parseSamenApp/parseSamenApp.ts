import ts, { ClassDeclaration } from "typescript"
import { ParseError } from "../errors"
import { ParsedError } from "../extractErrors/parseThrowStatement"
import parseServiceDefinition from "./parseServiceDefinition"
import { hasModifier } from "../tsUtils"

export interface ParsedSamenApp {
  models: Model[]
  errors: ParsedError[]
  services: ParsedSamenServiceDefinition[]
}

export interface ParsedSamenServiceDefinition {
  name: string
  models: Model[]
  errors: ParsedError[]
  funcs: ParsedSamenFunctionDefinition[]
  config: ParsedSamenServiceConfig
}

export type Model =
  | ts.InterfaceDeclaration
  | ts.TypeAliasDeclaration
  | ts.EnumDeclaration

export interface ParsedSamenFunctionDefinition {
  name: string
  actualFunction: ts.FunctionLikeDeclarationBase
  parameters: ts.ParameterDeclaration[]
  returnType: ts.TypeNode
  serviceContext?: {
    type: ts.TypeNode
    paramName?: string
  }
}

export interface ParsedSamenServiceConfig {
  middleware?: ParsedMiddlewareConfig[]
  contextType?: ts.TypeNode
}

export interface ParsedMiddlewareConfig {
  paramsType: ts.TypeNode
  nextType: ts.TypeNode | undefined
  contextType: ts.TypeNode
  middleware: ts.FunctionLikeDeclarationBase
}

export function parseSamenApp(
  samenSourceFile: ts.SourceFile,
  typeChecker: ts.TypeChecker,
): ParsedSamenApp {
  const exportStatements = samenSourceFile.statements.filter(
    (s) =>
      hasModifier(s, ts.SyntaxKind.ExportKeyword) || ts.isExportDeclaration(s),
  )

  const services: ParsedSamenServiceDefinition[] = []

  for (const statement of exportStatements) {
    if (ts.isVariableStatement(statement)) {
      for (const varDeclr of statement.declarationList.declarations) {
        const service = parseServiceDefinition(varDeclr, typeChecker)
        services.push(service)
      }
    } else if (ts.isExportDeclaration(statement)) {
      if (!statement.exportClause) {
        throw new ParseError(
          `"export * from './file'" are not supported`,
          statement,
        )
      } else if (!ts.isNamedExports(statement.exportClause)) {
        throw new ParseError("Unsupported export statement", statement)
      }

      for (const specifier of statement.exportClause.elements) {
        const service = parseServiceDefinition(specifier, typeChecker)
        services.push(service)
      }
    } else {
      throw new ParseError("Unsupported export statement", statement)
    }
  }

  const modelMap: Map<string, Model> = new Map<string, Model>()
  const errorMap: Map<string, ParsedError> = new Map<string, ParsedError>()

  for (const service of services) {
    for (const model of service.models) {
      const modelName = model.name.text
      if (!modelMap.has(modelName)) {
        modelMap.set(modelName, model)
      } else if (modelMap.get(modelName) !== model) {
        throw new ParseError(
          "You already have a different model with the same name, currently this is not possible. We intent to implement namespaces soon, stay tuned.",
          model,
        )
      }
    }
    for (const parsedError of service.errors) {
      const errorName = parsedError.name
      if (!errorMap.has(errorName)) {
        errorMap.set(errorName, parsedError)
      } else if (errorMap.get(errorName)?.ref !== parsedError.ref) {
        throw new ParseError(
          "You already have a different error class with the same name, currently this is not possible. We intent to implement namespaces soon, stay tuned.",
          parsedError.ref,
        )
      }
    }
  }

  return {
    models: [...modelMap.values()],
    errors: [...errorMap.values()],
    services,
  }
}

function removeShared<T>(objs: T[], sharedObjs: T[]): T[] {
  return objs.filter((m) => !sharedObjs.includes(m))
}

function deduplicate<T>(objs: T[]): T[] {
  return objs.reduce(
    (result, obj) => (result.includes(obj) ? result : [...result, obj]),
    [] as T[],
  )
}
