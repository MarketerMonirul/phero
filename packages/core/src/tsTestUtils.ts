import ts, {
  FunctionLikeDeclarationBase,
  VariableDeclaration,
} from "typescript"
import { TSFiles, VirtualCompilerHost } from "./VirtualCompilerHost"
import { PheroApp } from "./parsePheroApp/domain"
import { KindToNodeMappings } from "./tsUtils"

export function printPheroApp(app: PheroApp): string {
  return JSON.stringify(
    {
      services: app.services.map((service) => ({
        name: service.name,
        funcs: service.funcs.map((func) => ({
          name: func.name,
          func: printFunctionDeclaration(func.ref),
        })),
      })),
    },
    null,
    4,
  )

  function printFunctionDeclaration(
    func: FunctionLikeDeclarationBase | VariableDeclaration,
  ): string {
    const funcName = func.name?.getText()

    if (!funcName) {
      throw new Error("Func must have name")
    }

    return func.kind === ts.SyntaxKind.FunctionDeclaration
      ? `[FunctionDeclaration(${funcName})]`
      : func.kind === ts.SyntaxKind.VariableDeclaration
      ? `[VariableDeclaration(${funcName})]`
      : func.kind === ts.SyntaxKind.FunctionExpression
      ? `[FunctionExpression(${funcName})]`
      : func.kind === ts.SyntaxKind.ArrowFunction
      ? `[ArrowFunction(${funcName})]`
      : "[UNKNOWN]"
  }
}

export function createTestProgram(input: TSFiles | string): ts.Program {
  const vHost = new VirtualCompilerHost()

  if (typeof input === "string") {
    vHost.addFile(`phero.ts`, input)
  } else {
    for (const [fileName, content] of Object.entries(input)) {
      vHost.addFile(`${fileName}.ts`, content)
    }
  }

  const program = vHost.createProgram("phero.ts")
  return program
}

export function compileStatement<SK extends ts.SyntaxKind>(
  code: string,
  syntaxKind: SK,
): { statement: KindToNodeMappings[SK]; typeChecker: ts.TypeChecker } {
  const prog = createTestProgram(code)
  const statements = prog.getSourceFile("phero.ts")?.statements
  if (statements?.length !== 1) {
    throw new Error("Should provide exactly 1 statement")
  }
  const statement = statements[0]
  if (statement.kind !== syntaxKind) {
    throw new Error(
      `SytaxKind of statement is ${statement.kind}, but ${syntaxKind} was expected`,
    )
  }

  return {
    typeChecker: prog.getTypeChecker(),
    statement: statement as KindToNodeMappings[SK],
  }
}

export function compileStatements(code: string): {
  statements: ts.Statement[]
  typeChecker: ts.TypeChecker
} {
  const prog = createTestProgram(code)
  const statements = prog.getSourceFile("phero.ts")?.statements
  if (!statements || statements.length < 1) {
    throw new Error("Should provide at least 1 statement")
  }

  return {
    typeChecker: prog.getTypeChecker(),
    statements: statements.map((s) => s),
  }
}

export function compileProgram(input: TSFiles): {
  statements: ts.Statement[]
  typeChecker: ts.TypeChecker
} {
  const prog = createTestProgram(input)
  const statements = prog.getSourceFile("phero.ts")?.statements
  if (!statements || statements.length < 1) {
    throw new Error("Should provide at least 1 statement")
  }

  return {
    typeChecker: prog.getTypeChecker(),
    statements: statements.map((s) => s),
  }
}

const printer = ts.createPrinter({
  newLine: ts.NewLineKind.LineFeed,
  noEmitHelpers: true,
  removeComments: true,
  omitTrailingSemicolon: false,
})

export function printCode(node: ts.Node): string {
  const sf = ts.createSourceFile(
    "a.ts",
    "",
    ts.ScriptTarget.ESNext,
    undefined,
    ts.ScriptKind.TS,
  )

  return printer.printNode(ts.EmitHint.Unspecified, node, sf)
}
