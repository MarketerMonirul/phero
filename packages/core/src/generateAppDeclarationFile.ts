import ts from "typescript"
import {
  generateErrorClass,
  generateFunction,
  generateModel,
  generateNamespace,
  ReferenceMaker,
} from "./code-gen-lib"
import { ParsedSamenApp } from "./parseSamenApp"
import { VirtualCompilerHost } from "./VirtualCompilerHost"
import * as tsx from "./tsx"

export default function generateAppDeclarationFile(
  app: ParsedSamenApp,
  typeChecker: ts.TypeChecker,
): string {
  const t1 = Date.now()

  const domainIdentifier = ts.factory.createIdentifier("domain")
  const versionIdentifier = ts.factory.createIdentifier("v_1_0_0")

  const refMaker = new ReferenceMaker(
    app.models,
    typeChecker,
    ts.factory.createQualifiedName(domainIdentifier, versionIdentifier),
  )

  const namespaceDeclrs: ts.ModuleDeclaration[] = []

  if (app.models.length || app.errors.length) {
    namespaceDeclrs.push(
      // export namespace domain {
      generateNamespace(domainIdentifier, [
        // export namespace v_1_0_0 {
        generateNamespace(versionIdentifier, [
          ...app.models.map((m) =>
            // export interface MyModel {
            generateModel(m, refMaker),
          ),
          ...app.errors.map((e) =>
            // export class Error {
            generateErrorClass(e, refMaker),
          ),
        ]),
      ]),
    )
  }

  const isUsingSamenContext = app.services.some((s) => !!s.config.contextType)
  if (isUsingSamenContext) {
    namespaceDeclrs.push(
      generateNamespace(ts.factory.createIdentifier("samen"), [
        tsx.typeAlias({
          export: true,
          name: "SamenContext",
          typeParameters: [tsx.typeParam({ name: "T" })],
          type: tsx.type.reference({ name: "T" }),
        }),
      ]),
    )
  }

  for (const service of app.services) {
    namespaceDeclrs.push(
      // export namespace cmsService {
      generateNamespace(ts.factory.createIdentifier(service.name), [
        // export namespace v_1_0_0 {
        generateNamespace(versionIdentifier, [
          // export function myFunction(): Promise<void> {
          ...service.funcs.map((func) => generateFunction(func, refMaker)),
        ]),
      ]),
    )
  }

  const vHost = new VirtualCompilerHost({
    emitDeclarationOnly: true,
  })

  vHost.addFile("api.ts", generateTS(namespaceDeclrs))

  const program = vHost.createProgram("api.ts")
  const emitResult = program.emit()

  if (emitResult.diagnostics.length) {
    console.error("OOPS SOMETHING IS WRONG")
    console.error(generateTS(namespaceDeclrs))
    console.error(emitResult)
  }

  const t2 = Date.now()
  // console.log("generateAppDeclarationFile in", t2 - t1)

  const declrFile = vHost.getFile("api.d.ts")

  if (!declrFile) {
    throw new Error("Can't generate app declaration file")
  }

  return declrFile
}

function generateTS(nodes: ts.Node[]): string {
  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
    removeComments: true,
  })

  const file = ts.createSourceFile(
    "api.ts",
    "",
    ts.ScriptTarget.ES5,
    false,
    ts.ScriptKind.TS,
  )

  return printer.printList(
    ts.ListFormat.SourceFileStatements,
    ts.factory.createNodeArray(nodes),
    file,
  )
}
