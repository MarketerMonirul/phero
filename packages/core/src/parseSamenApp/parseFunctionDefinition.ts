import ts from "typescript"
import { ParseError } from "../errors"
import parseReturnType from "./parseReturnType"
import { ParsedSamenFunctionDefinition } from "./parseSamenApp"
import { resolveSymbol } from "../tsUtils"

export default function parseFunctionDefinition(
  node: ts.ObjectLiteralElementLike | ts.VariableDeclaration,
  typeChecker: ts.TypeChecker,
): ParsedSamenFunctionDefinition {
  if (ts.isSpreadAssignment(node)) {
    throw new ParseError(
      "S116: Sorry, no support for spread assignment (yet)",
      node,
    )
  }

  const parsedSamenFunctionDef: ParsedSamenFunctionDefinition = {
    name: parseFunctionName(node.name),
    ...parseActualFunction(node, typeChecker),
  }

  return parsedSamenFunctionDef
}

function parseFunctionName(
  functionName:
    | ts.Identifier
    | ts.StringLiteral
    | ts.NumericLiteral
    | ts.ComputedPropertyName
    | ts.PrivateIdentifier
    | ts.ObjectBindingPattern
    | ts.ArrayBindingPattern,
): string {
  if (
    ts.isNumericLiteral(functionName) ||
    ts.isComputedPropertyName(functionName) ||
    ts.isPrivateIdentifier(functionName) ||
    ts.isObjectBindingPattern(functionName) ||
    ts.isArrayBindingPattern(functionName)
  ) {
    throw new ParseError(
      "S117: Function name should have a clear identifier, no support for computed names or binding patterns" +
        functionName.kind,
      functionName,
    )
  }
  return functionName.text
}

function parseActualFunction(
  node: ts.Node,
  typeChecker: ts.TypeChecker,
): Pick<
  ParsedSamenFunctionDefinition,
  "actualFunction" | "parameters" | "returnType"
> {
  if (ts.isShorthandPropertyAssignment(node)) {
    const symbol = typeChecker.getShorthandAssignmentValueSymbol(node)

    if (!symbol || !symbol.declarations?.[0]) {
      throw new ParseError(`S118: Can't find function (${node.kind})`, node)
    }

    return parseActualFunction(symbol.declarations?.[0], typeChecker)
  }

  if (ts.isImportSpecifier(node)) {
    return parseActualFunction(node.name, typeChecker)
  }

  if (ts.isPropertyAssignment(node)) {
    return parseActualFunction(node.initializer, typeChecker)
  }

  if (ts.isFunctionExpression(node) || ts.isArrowFunction(node)) {
    return {
      actualFunction: node,
      parameters: node.parameters.map((p) => p),
      returnType: parseReturnType(node),
    }
  }

  if (ts.isFunctionDeclaration(node)) {
    return {
      actualFunction: node,
      parameters: node.parameters.map((p) => p),
      returnType: parseReturnType(node),
    }
  }

  if (ts.isVariableDeclaration(node) && node.initializer) {
    if (ts.isArrowFunction(node.initializer)) {
      return {
        actualFunction: node.initializer,
        parameters: node.initializer.parameters.map((p) => p),
        returnType: parseReturnType(node.initializer),
      }
    }
    return parseActualFunction(node.initializer, typeChecker)
  }

  if (ts.isIdentifier(node)) {
    const symbol = resolveSymbol(node, typeChecker)
    if (!symbol?.declarations?.[0]) {
      throw new ParseError(`S119: Can't find function [${node.kind}]`, node)
    }
    return parseActualFunction(symbol.declarations?.[0], typeChecker)
  }

  if (ts.isPropertyAccessExpression(node)) {
    const lastToken = node.getLastToken()
    if (lastToken) {
      return parseActualFunction(lastToken, typeChecker)
    }
  }

  throw new ParseError("S120: Unsupported syntax" + node.kind, node)
}
