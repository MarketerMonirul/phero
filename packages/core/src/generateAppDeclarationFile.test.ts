import ts from "typescript"
import generateAppDeclarationFile from "./generateAppDeclarationFile"

import parseSamenApp from "./parseSamenApp"
import { createTestProgram } from "./tsTestUtils"

function generate(prog: ts.Program): string {
  // if (prog.getSemanticDiagnostics().length) {
  //   // console.log("OEPS COMPILE ERRORS DETECTED")
  //   console.error(prog.getSemanticDiagnostics())
  //   throw new Error("Compile error")
  // }
  const typeChecker = prog.getTypeChecker()
  const parsedApp = parseSamenApp(prog.getSourceFile("samen.ts")!, typeChecker)
  const dts = generateAppDeclarationFile(parsedApp, typeChecker)
  return dts
}

describe("generateAppDeclarationFile", () => {
  describe("middleware", () => {
    test("should parse middleware", () => {
      const parsedApp = generate(
        createTestProgram(`
        type NextFunction<T = void> = T extends void
          ? () => Promise<void>
          : (ctx: T) => Promise<void>
        type SamenContext<T = {}> = T
        type SamenParams<T = {}> = Partial<T>

        async function getArticle(aap: string, ctx: SamenContext<{ user: User }>): Promise<string> {
          return "ok"
        }

        interface User { uid: string }

        async function requireUID(params: SamenParams, ctx: SamenContext<{ uid: string }>, next: NextFunction<{ uid: string, x: string }>) {
          // valideer id token
          await next({ uid, x })
        }
        
        async function requireCmsUser(params: SamenParams, ctx: SamenContext<{ uid: string, x: string }>, next: NextFunction<{ user: User }>) {
          // zet om naar user, of het cms user is
          await next({ user: { uid } })
        }

        export const articleService = createService({
          getArticle: createFunction(getArticle),
        }, {
          middleware: [requireUID, requireCmsUser],
        })
      `),
      )

      // EXPECT function getArticle(aap: string): Promise<string>;
      // because x:number is provided implicitly by middleware
      expect(parsedApp).toMatchSnapshot()
    })
  })
  describe("errors", () => {
    test("should generate service error", () => {
      const parsedApp = generate(
        createTestProgram(`
          class ArticleError extends Error {
          }

          async function getArticle(): Promise<string> {
            throw new ArticleError()
          }

          export const articleService = createService({
            getArticle: createFunction(getArticle),
          })
        `),
      )

      expect(parsedApp).toMatchSnapshot()
    })

    test("should generate multi-service multi-error", () => {
      const parsedApp = generate(
        createTestProgram(`
          class HamburgerError extends Error {
          }

          async function getHamburger(): Promise<string> {
            throw new HamburgerError()
          }
          

          export const hamburgerService = createService({
            getHamburger: createFunction(getHamburger),
          })
          
          class ArticleError extends Error {
          }

          async function getArticle(): Promise<string> {
            throw new ArticleError()
          }

          export const articleService = createService({
            getArticle: createFunction(getArticle),
          })
        `),
      )

      expect(parsedApp).toMatchSnapshot()
    })

    test("should generate common errors in domain", () => {
      const parsedApp = generate(
        createTestProgram(`
          class CommonError extends Error {
          }

          class HamburgerError extends Error {
          }

          async function getHamburger(): Promise<string> {
            throw new HamburgerError()
          }

          async function setHamburger(): Promise<string> {
            throw new CommonError()
          }
          
          export const hamburgerService = createService({
            getHamburger: createFunction(getHamburger),
            setHamburger: createFunction(setHamburger),
          })
          
          class ArticleError extends Error {
          }

          async function getArticle(): Promise<string> {
            throw new ArticleError()
          }

          async function setArticle(): Promise<string> {
            throw new CommonError()
          }

          export const articleService = createService({
            getArticle: createFunction(getArticle),
            setArticle: createFunction(setArticle),
          })
        `),
      )

      expect(parsedApp).toMatchSnapshot()
    })

    test("should generate properties on generated error classes", () => {
      const parsedApp = generate(
        createTestProgram(`
          class GenericError extends Error {
            constructor(test: number) {
              super(\`Test \${test}\`)
            }
          }

          class ArticleError extends GenericError {
            public myPublicProp: number
            
            constructor(public myProp: number, myParam: string) {
              super(myProp + 1)
              this.myPublicProp = myProp + 2
            }
          }

          async function getArticle(): Promise<string> {
            throw new ArticleError(123)
          }

          export const articleService = createService({
            getArticle: createFunction(getArticle),
          })
        `),
      )

      expect(parsedApp).toMatchSnapshot()
    })
  })

  describe("typeAliases", () => {
    test("type aliases which are aliases of aliases", () => {
      const parsedApp = generate(
        createTestProgram({
          routine: `
          export async function getRoutine(versionId: string): Promise<Result> {
            return {x: 1}
          }

          export type Result = TypeRef

          export type TypeRef = TypeReal

          export type TypeReal = {x: number}
          `,

          samen: `
          import {getRoutine} from './routine'
          
          export const workoutRoutineService = createService({
            getRoutine: createFunction(getRoutine),
          })
        `,
        }),
      )

      expect(parsedApp).toMatchSnapshot()
    })
  })
})
