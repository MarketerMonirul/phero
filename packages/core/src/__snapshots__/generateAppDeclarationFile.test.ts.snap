// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateAppDeclarationFile errors should generate common errors in domain 1`] = `
"export declare namespace domain {
    namespace v_1_0_0 {
        class HamburgerError extends Error {
            constructor(message: string);
        }
        class CommonError extends Error {
            constructor(message: string);
        }
        class ArticleError extends Error {
            constructor(message: string);
        }
    }
}
export declare namespace hamburgerService {
    namespace v_1_0_0 {
        function getHamburger(): Promise<string>;
        function setHamburger(): Promise<string>;
    }
}
export declare namespace articleService {
    namespace v_1_0_0 {
        function getArticle(): Promise<string>;
        function setArticle(): Promise<string>;
    }
}
"
`;

exports[`generateAppDeclarationFile errors should generate multi-service multi-error 1`] = `
"export declare namespace domain {
    namespace v_1_0_0 {
        class HamburgerError extends Error {
            constructor(message: string);
        }
        class ArticleError extends Error {
            constructor(message: string);
        }
    }
}
export declare namespace hamburgerService {
    namespace v_1_0_0 {
        function getHamburger(): Promise<string>;
    }
}
export declare namespace articleService {
    namespace v_1_0_0 {
        function getArticle(): Promise<string>;
    }
}
"
`;

exports[`generateAppDeclarationFile errors should generate properties on generated error classes 1`] = `
"export declare namespace domain {
    namespace v_1_0_0 {
        class ArticleError extends Error {
            constructor(message: string, myPublicProp: number, myProp: number);
        }
    }
}
export declare namespace articleService {
    namespace v_1_0_0 {
        function getArticle(): Promise<string>;
    }
}
"
`;

exports[`generateAppDeclarationFile errors should generate service error 1`] = `
"export declare namespace domain {
    namespace v_1_0_0 {
        class ArticleError extends Error {
            constructor(message: string);
        }
    }
}
export declare namespace articleService {
    namespace v_1_0_0 {
        function getArticle(): Promise<string>;
    }
}
"
`;

exports[`generateAppDeclarationFile middleware should parse middleware 1`] = `
"export declare namespace domain {
    namespace v_1_0_0 {
        interface User {
            uid: string;
        }
    }
}
export declare namespace samen {
    type SamenContext<T> = T;
}
export declare namespace articleService {
    namespace v_1_0_0 {
        function getArticle(ctx: samen.SamenContext<{
            uid: string;
        }>, aap: string): Promise<string>;
    }
}
"
`;

exports[`generateAppDeclarationFile typeAliases type aliases to native array types 1`] = `
"export declare namespace domain {
    namespace v_1_0_0 {
        type AadReturnType1 = string[];
    }
}
export declare namespace articleService {
    namespace v_1_0_0 {
        function getArticle(): Promise<domain.v_1_0_0.AadReturnType1>;
    }
}
"
`;

exports[`generateAppDeclarationFile typeAliases type aliases to native types 1`] = `
"export declare namespace domain {
    namespace v_1_0_0 {
        type AadReturnType2 = boolean;
    }
}
export declare namespace articleService {
    namespace v_1_0_0 {
        function getArticle(): Promise<{
            aad: domain.v_1_0_0.AadReturnType2;
        }>;
    }
}
"
`;

exports[`generateAppDeclarationFile typeAliases type aliases which are aliases of aliases 1`] = `
"export declare namespace domain {
    namespace v_1_0_0 {
        type TypeReal = {
            x: number;
        };
        type TypeRef = domain.v_1_0_0.TypeReal;
        type Result = domain.v_1_0_0.TypeRef;
    }
}
export declare namespace workoutRoutineService {
    namespace v_1_0_0 {
        function getRoutine(versionId: string): Promise<domain.v_1_0_0.Result>;
    }
}
"
`;
